### ISSUE TITLE
MemCacheProtocol does not return an error if it is used after disconnecting

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3643]
|<img alt="viiru's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| viiru reported|
|-|-|
|Trac ID|trac#3643|
|Type|defect|
|Created|2009-02-02 12:52:43Z|
|Branch|https://github.com/twisted/twisted/tree/disconnected-memcache-3643-3|

If I connect to a memcached using the MemCacheProtocol and use it's set or get methods after the server has exited or otherwise the connection has been lost the protocol goes on happily generating deferreds that never return.

Tested with version 8.1.0-4 as packaged in Debian Lenny.

<details><summary>Searchable metadata</summary>

```
trac-id__3643 3643
type__defect defect
reporter__viiru viiru
priority__high high
milestone__ 
branch__branches_disconnected_memcache_3643_3 branches/disconnected-memcache-3643-3
branch_author__therve__washort therve, washort
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1233579163000000 1233579163000000
changetime__1252344893000000 1252344893000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_1"></a>|@washort commented|
|-|-|

(In [26247]) Branching to 'disconnected-memcache-3643'

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_2"></a>|@washort removed owner|
|-|-|

memcache command methods now fail when not called on a connected protocol.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_3"></a>|@washort commented|
|-|-|

(In [26251]) Branching to 'ctcp-version-3643'

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_5"></a>|@mwhudson set owner to @washort|
|-|-|

Doesn't this approach still have the problem that if you make a request and the connection is lost before you get a reply, the deferred that you got will never fire?  Maybe that's OK.

This branch is still an improvement.  The only comment I have on the code is that CommandFailureTests._test should have a new docstring that explains how it's different from the other _test.

So, please think about the question above, and if it's OK tweak the docstring and land.  Otherwise, more coding awaits!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

> Doesn't this approach still have the problem that if you make a request and the connection is lost before you get a reply, the deferred that you got will never fire? Maybe that's OK. 

That sounds _really_ not okay.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_7"></a>|@washort removed owner|
|-|-|

Added a test for disconnection behaviour.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_8"></a>|@jerub set owner to @washort|
|-|-|

```twisted.test.test_memcache.MemCacheTestCase.test_connectionLost``` does not actually test that connectionLost causes a failure, as there isn't a callback, only an errback. If cmd.success() is called instead of fail(), the test passes.

I have checked that the test coverage hits all the data, which is great.

```twisted.protocols.memcache``` line 174 has 11 space indent instead of 12. It's a niggle, it's not in your code, but it's something that you should fix in this branch while you're here.

Good work :)

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_9"></a>|@washort removed owner|
|-|-|

Fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @washort|
|-|-|

```
exarkun@boson:~/Projects/Twisted/trunk$ grep connected twisted/internet/interfaces.py
        If we are connected, disconnect, if we are trying to connect,
        Called every time this is connected to a Port or Connector.
        Called every time this is unconnected from a Port or Connector.
        Close a file descriptor which is connected to the child process, identified
        @param addr: a tuple of (ip, port). For connected transports must
                     be the address the transport is connected to, or None.
                     In non-connected mode this is mandatory.
        This changes it to connected mode. Datagrams can only be sent to
        Write packet to address we are connected to.
        Write packet to address we are connected to.
```

(Or to be more direct: there is no documented `connected` attribute on transports; try adding something to `connectionLost` to track this state.)

In `CommandFailureTests`, I think you're looking for `assertFailure`.  As written, `_test` on that class doesn't verify that the Deferred fails; it only verifies that _if_ it fails, it does so with a `RuntimeError`.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

(In [27162]) Branching to 'disconnected-memcache-3643-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve removed owner|
|-|-|

Please review!

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_14"></a>|@mwhudson set owner to @therve|
|-|-|

Not a full review, but: it would be nice if CommandMixin had a _test method that raised NotImplementedError and had a docstring that explained what was expected of it.

It would be *really* nice if CommandFailureTests._test's docstring wasn't the same as the other one.  And that docstring sucks too -- it's a "what docstring" rather than a "why docstring", and not even a very good one at that.

Otherwise it looks basically fine but I don't have time to read it really closely now.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve commented|
|-|-|

(In [27179]) Branching to 'disconnected-memcache-3643-3'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve removed owner|
|-|-|

I tried to improve the docstrings, I hope it will be bit better now.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_17"></a>|@jonathanj set owner to @therve|
|-|-|

Some documentation issues:

 1. A bunch of the tests have docstrings of the form "Test that X does Y." Since you're changing/adding these anyway, it would probably be a nice idea to rephrase them to exclude the "Test that" part.
 2. `True` is not the type of `_disconnected`

Apart from those, the patch and the coverage looks good, so go ahead and merge.

**Comment by twisted-trac** [comment]
|<img alt="powdahound's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|powdahound commented|
|-|-|

Anything I can do to help move this ticket along? I'd like to make use of the fix in trunk.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve removed owner|
|-|-|

I discovered some bugs in a test. It would need another review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to @therve|
|-|-|

  1. `CommandFailureTests.timeout` seems superfluous.
  1. Maybe it would be good to have a direct unit test for the fact that the code needs to do `while` instead of `for` in the places where it's doing that.
  1. There's two loops over `self._current`.  Can you refactor so it's just one?

Otherwise, seems good.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve removed owner|
|-|-|

Thanks for the review! Handled, I think.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve commented|
|-|-|

(In [27292]) Handle review comments

Refs [#3643](https://github.com/twisted/twisted/issues/3643)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_23"></a>|@jonathanj set owner to @therve|
|-|-|

I think this looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve set status to `closed`|
|-|-|

(In [27301]) Merge disconnected-memcache-3643-3 

Authors: dash, therve
Reviewers: exarkun, jonathanj, mwh
Fixes [#3643](https://github.com/twisted/twisted/issues/3643)

Make MemcacheProtocol methods fail when the underlying transport is
disconnected: previously it would let the commands state forever.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
